curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+    uint64_t readLen = 0;
+    int ret = 0;
+    uint64_t readOffset = 0;
+    std::vector<ReadRequest> totalRequests;
+    while (length > 0) {
+        std::vector<ReadRequest> requests;
+        if (chunkPos + length > chunkSize) {
+            readLen = chunkSize - chunkPos;
+        } else {
+            readLen = length;
+        }
+
+        ReadChunk(index, chunkPos, readLen, dataBuf + readOffset, readOffset,
+                  &requests);
+        totalRequests.insert(totalRequests.end(), requests.begin(),
+                             requests.end());
+        length -= readLen;
+        index++;
+        readOffset += readLen;
+        chunkPos = (chunkPos + readLen) % chunkSize;
+    }
+
+    std::vector<S3ReadRequest> totalS3Requests;
+    auto iter = totalRequests.begin();
+
+    for (; iter != totalRequests.end(); iter++) {
+        auto s3InfoListIter = inode->s3chunkinfomap().find(iter->index);
+        if (s3InfoListIter == inode->s3chunkinfomap().end()) {
+            memset(dataBuf + iter->bufOffset, 0, iter->len);
+            continue;
+        }
+        std::vector<S3ReadRequest> s3Requests;
+        GenerateS3Request(*iter, s3InfoListIter->second, dataBuf, &s3Requests);
+        totalS3Requests.insert(totalS3Requests.end(), s3Requests.begin(),
+                               s3Requests.end());
+    }
+    uint32_t i;
+    for (i = 0; i < totalS3Requests.size(); i++) {
+        S3ReadRequest tmp_req = totalS3Requests[i];
+        LOG(INFO) << "S3ReadRequest readoffset:" << tmp_req.GetReadOffset()
+                  << ",offset:" << tmp_req.GetS3ChunkInfo().offset()
+                  << ",len:" << tmp_req.GetS3ChunkInfo().len();
+    }
+
+    std::vector<S3ReadResponse> responses;
+
+    ret = handleReadRequest(totalS3Requests, &responses);
+    if (ret < 0) {
+        LOG(ERROR) << "handle read request fail:" << ret;
+        return ret;
+    }
+
+    auto repIter = responses.begin();
+    for (; repIter != responses.end(); repIter++) {
+        LOG(INFO) << "readOffset:" << repIter->GetReadOffset()
+                  << ",bufLen:" << repIter->GetBufLen();
+        strncpy(dataBuf + repIter->GetReadOffset(), repIter->GetDataBuf(),
+                repIter->GetBufLen());
+    }
+
+    return readOffset;
+}
+
+int FileCacheManager::handleReadRequest(
+    const std::vector<S3ReadRequest>& requests,
+    std::vector<S3ReadResponse>* responses) {
+    uint64_t chunkSize = s3ClientAdaptor_->GetChunkSize();
+    uint64_t blockSize = s3ClientAdaptor_->GetBlockSize();
+    (*responses).reserve(requests.size());
+    std::vector<S3ReadRequest>::const_iterator iter = requests.begin();
+    std::atomic<uint64_t> pendingReq(0);
+    curve::common::CountDownEvent cond(1);
+    bool async = false;
+    GetObjectAsyncCallBack cb =
+        [&](const S3Adapter* adapter,
+            const std::shared_ptr<GetObjectAsyncContext>& context) {
+            pendingReq.fetch_sub(1, std::memory_order_relaxed);
```

Comment:
async callback should judge return code of the rpc call.

Reply:
todo, async is not realized